doc/man/persist-tool.1.in
'\" t
.\"     Title: persist-tool
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 07/05/2016
.\"    Manual: The persist-tool tool manual page
.\"    Source: 5 F6
.\"  Language: English
.\"
.TH "PERSIST\-TOOL" "1" "07/05/2016" "5 F6" "The persist-tool tool manual p"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
persist-tool \- Display the content of the persist file
.SH "SYNOPSIS"
.HP \w'\fBpersist\-tool\fR\ 'u
\fBpersist\-tool\fR [command] [options]
.SH "DESCRIPTION"
.PP
NOTE: The persist\-tool application is distributed with the system logging application, and is usually part of the syslog\-ng package\&. The latest version of the syslog\-ng application is available at the
\m[blue]\fBofficial syslog\-ng website\fR\m[]\&.
.PP
This manual page is only an abstract; for the complete documentation of syslog\-ng, see
\m[blue]\fBThe Administrator Guide\fR\m[]\&.
.PP
The
\fBpersist\-tool\fR
application is a utility that can be used to dump the content of the persist file, and manipulate its content\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBWarning\fR
.ps -1
.br
.PP
Persist\-tool is a special tool for syslog\-ng experts\&. Do use the tool unless you know exactly what you are doing\&. Misconfiguring it will result in irrecoverable damage to the persist file, without any warning\&.
.sp .5v
.RE
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
Limitations:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
The persist\-state functions can be used only with syslog\-ng PE 5 LTS style persist file (SLP4)\&. Older persist files are not supported\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Wildcard characters are not supported in file/directory names\&.
.RE
.sp .5v
.RE
.SH "THE DUMP COMMAND"
.HP \w'\fBdump\fR\ 'u
\fBdump\fR [options] [persist_file]
.PP
Use the
\fBdump\fR
command to print the current content of the persist file in JSON format to the console\&.
.PP
The
\fBdump\fR
command has the following options:
.PP
\fB\-\-help\fR or \fB\-?\fR
.RS 4
Display a brief help message\&.
.RE
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
persist\-tool dump /opt/syslog\-ng/var/syslog\-ng\&.persist 
.fi
.if n \{\
.RE
.\}
.PP
The output looks like:
.sp
.if n \{\
.RS 4
.\}
.nf
evtxlog_reader_curpos(Application) = { "version": 0, "big_endian": false, "bookmark_xml": "<BookmarkList Direction=\*(Aqbackward\*(Aq>\er\en  <Bookmark Channel=\*(AqApplication\*(Aq RecordId=\*(Aq2610170\*(Aq IsCurrent=\*(Aqtrue\*(Aq\e
/>\er\en<\e/BookmarkList>" }
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
affile_sd_curpos(/tmp/foo\&.txt) = { "version": 1, "big_endian": false, "raw_buffer_leftover_size": 0, "buffer_pos": 0, "pending_buffer_end": 0, "buffer_size
": 0, "buffer_cached_eol": 0, "pending_buffer_pos": 0, "raw_stream_pos": 64, "pending_raw_stream_pos": 0, "raw_buffer_size": 0, "pending_raw_buffer_size": 0, "file_size": 64, "file_inode": 0, "run_id"
: 0 }
.fi
.if n \{\
.RE
.\}
.SH "THE GENERATE COMMAND"
.HP \w'\fBgenerate\fR\ 'u
\fBgenerate\fR [options]
.PP
Use the
\fBgenerate\fR
command to generate a persist file based on the content of the given configuration file\&. It will find all files and event sources that match the sources configured in the configuration file, and generate a state entry into the persist file for all of them\&. The position for the state entries will be set to the end of the source, therefore syslog\-ng will continue them from the last position instead of sending messages from the beginning\&. Any other type of entries will be skipped from persist generation\&. If the persist file already exists, persist\-tool will display an error message and exit to prevent overwriting an already existing persist file\&.
.PP
To use the
\fBgenerate\fR
command: stop syslog\-ng, remove the persist file, then execute
\fBpersist\-tool generate\fR\&.
.PP
The
\fBgenerate\fR
command has the following options:
.PP
\fB\-\-help\fR or \fB\-?\fR
.RS 4
Display a brief help message\&.
.RE
.PP
\fB\-\-config\-file=<config_file>\fR or \fB\-c\fR
.RS 4
Required parameter\&. The syslog\-ng configuration file that will be the base of the generated persist file\&.
.RE
.PP
\fB\-\-force\fR or \fB\-f\fR
.RS 4
Overwrite the persist file if it already exists\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBWarning\fR
.ps -1
.br
Use this option with care, persist\-tool will not ask for confirmation\&.
.sp .5v
.RE
.RE
.PP
\fB\-\-output\-dir=<directory>\fR or \fB\-o\fR
.RS 4
Required parameter\&. The directory where the persist file will be generated to\&. The name of the persist file will be
syslog\-ng\&.persist\&.
.RE
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
/opt/syslog\-ng/bin/persist\-tool generate \-c /opt/syslog\-ng/etc/syslog\-ng\&.conf \-o \&.
.fi
.if n \{\
.RE
.\}
.PP
The output looks like:
.sp
.if n \{\
.RS 4
.\}
.nf
New persist file generated: \&./syslog\-ng\&.persist
.fi
.if n \{\
.RE
.\}
.SH "THE GENERATE-AGENT COMMAND"
.HP \w'\fBgenerate\-agent\fR\ 'u
\fBgenerate\-agent\fR [options]
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
This command is only available on Windows\&.
.sp .5v
.RE
.PP
Use the
\fBgenerate\-agent\fR
command to generate a persist file based on the registry by default, or use the content of the given XML configuration file\&. It will find all files and event sources that match the sources configured in the registry or configuration file, and generate a state entry into the persist file for all of them\&. The position for the state entries will be set to the end of the source, therefore syslog\-ng will continue them from the last position instead of sending messages from the beginning\&. Any other type of source will be skipped from persist generation\&. If the persist file already exists, persist\-tool will display an error message and exit to prevent overwriting an already existing persist file\&.
.PP
To use the
\fBgenerate\fR
command: stop syslog\-ng, remove the persist file, then execute
\fBpersist\-tool generate\fR\&.
.PP
The
\fBgenerate\-agent\fR
command has the following options:
.PP
\fB\-\-help\fR or \fB\-?\fR
.RS 4
Display a brief help message\&.
.RE
.PP
\fB\-\-force\fR or \fB\-f\fR
.RS 4
Overwrite the persist file if it already exists\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBWarning\fR
.ps -1
.br
Use this option with care, persist\-tool will not ask for confirmation\&.
.sp .5v
.RE
.RE
.PP
\fB\-\-output\-dir=<directory>\fR or \fB\-o\fR
.RS 4
The directory where the persist file will be generated to\&. The name of the persist file will be
syslog\-ng\&.persist\&.
.RE
.PP
\fB\-\-xml\-config\fR or \fB\-x\fR
.RS 4
The XML configuration file of the syslog\-ng Agent for Windows application\&. If this parameter is set, the persist file will be generated from the XML configuration file\&. If this parameter is missing, the configuration available in the system registry will be used\&.
.RE
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
persist\-tool\&.exe generate\-agent \-f \-x c:\eeee\&.xml \-o "c:\eprogram files\esyslog\-ng agent\evar"
.fi
.if n \{\
.RE
.\}
.SH "THE ADD COMMAND"
.HP \w'\fBadd\fR\ 'u
\fBadd\fR [options] [input_file]
.PP
Use the
\fBadd\fR
command to add or modify a specified state\-entry in the persist file\&. The state\-entry should be in the same format as the
\fBdump\fR
command displays it\&. If the given state\-entry already exists, it will be updated\&. Otherwise, a new value will be added\&. If the given persist state is invalid, it will be skipped\&.
.PP
To use the
\fBadd\fR
command: use
\fBpersist\-tool dump\fR
to print the content of the current persist file, and redirect it to a file\&. Edit the content of this file\&. Use
\fBpersist\-tool add\fR
with this file to modify the persist\&.
.PP
The
\fBadd\fR
command has the following options:
.PP
\fB\-\-help\fR or \fB\-?\fR
.RS 4
Display a brief help message\&.
.RE
.PP
\fB\-\-output\-dir=<directory>\fR or \fB\-o\fR
.RS 4
Required parameter\&. The directory where the persist file is located at\&. The name of the persist file stored in this directory must be
syslog\-ng\&.persist\&.
.RE
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
/opt/syslog\-ng/bin/persist\-tool add dump_persist \-o \&.
.fi
.if n \{\
.RE
.\}
.PP
The valid output looks like:
.sp
.if n \{\
.RS 4
.\}
.nf
log_reader_curpos(Application)      OK
affile_sd_curpos(/var/aaa\&.txt)        OK
.fi
.if n \{\
.RE
.\}
.PP
The invalid output looks like:
.sp
.if n \{\
.RS 4
.\}
.nf
log_reader_curpos(Application)      OK
wrong
        FAILED (error: Invalid entry syntax)
affile_sd_curpos(/var/aaa\&.txt)        OK
.fi
.if n \{\
.RE
.\}
.SH "THE RENAME-KEY COMMAND"
.HP \w'\fBrename\-key\fR\ 'u
\fBrename\-key\fR [\-\-old_key\ <old\-key\-in\-persist\-file>] [\-\-new_key\ <new\-key\-in\-persist\-file>] [persist\-file]
.PP
Use the
\fBrename\-key\fR
command rename a specified state\-entry in the persist file\&. The key of the state\-entry to rename should be the same as the
\fBdump\fR
command displays it\&. After renaming, the
\fBdump\fR
command of the persist file shows both the old entry with the old key and a new entry with new key, but the old entry will be flagged as unused\&. When is started and loads the modified persist file, it automatically deletes the old entry\&.
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
persist\-tool rename\-key \-o "java,org\&.syslog_ng\&.hdfs\&.HdfsDestination" \-n "java_dst(hdfs,hdfs://10\&.140\&.32\&.80:8020,var/dest/hadoop\&.txt)" syslog\-ng\&.persist
.fi
.if n \{\
.RE
.\}
.SH "FILES"
.PP
/opt/syslog\-ng/bin/persist\-tool
.SH "SEE ALSO"
.PP
\fBsyslog\-ng\&.conf\fR(5)
.PP
\fBsyslog\-ng\fR(8)
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
For the detailed documentation of see
\m[blue]\fB\fBThe 5 F6 Administrator Guide\fR\fR\m[]
.PP
If you experience any problems or need help with syslog\-ng, visit
\m[blue]\fBvisit the syslog\-ng wiki\fR\m[]
or the
\m[blue]\fB\fBsyslog\-ng mailing list\fR\fR\m[]\&.
.PP
For news and notifications about of syslog\-ng, visit the
\m[blue]\fB\fBsyslog\-ng Blog\fR\fR\m[]\&.
.sp .5v
.RE
.SH "AUTHOR"
.PP
This manual page was written by the BalaBit Documentation Team <documentation@balabit\&.com>\&.
.SH "COPYRIGHT"


persist-tool/persist-tool.c
/*
 * Copyright (c) 2002-2010 Balabit
 * Copyright (c) 1998-2010 Balázs Scheidler
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * As an additional exemption you are allowed to compile & link against the
 * OpenSSL libraries as published by the OpenSSL project. See the file
 * COPYING for details.
 *
 */

#include "dump.h"
#include "add.h"
#include "generate.h"
#include "rename.h"
#include "generate-agent.h"
#include "persist-tool.h"
#include "mainloop.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#if HAVE_GETOPT_H
#include <getopt.h>
#endif

#ifndef _WIN32
#define STATE_HANDLER_MODULES "afsqlsource,basic-proto,rltp-proto,disk-buffer,eventlog"
#else
#define STATE_HANDLER_MODULES "basic-proto,rltp-proto,disk-buffer,eventlog"
#endif

static void
load_state_handler_modules(GlobalConfig *cfg)
{
  gint i;
  gchar **mods;
  mods = g_strsplit(STATE_HANDLER_MODULES, ",", 0);
  for (i = 0; mods[i]; i++)
    {
      plugin_load_module(mods[i], cfg, NULL );
    }
  g_strfreev(mods);
}

static gboolean
persist_tool_start_state(PersistTool *self)
{
  if (!g_threads_got_initialized)
    {
      g_thread_init(NULL);
    }
  main_thread_handle = get_thread_id();
  self->state = persist_state_new(self->persist_filename);
  persist_state_set_mode(self->state, self->mode);
  if (!persist_state_start(self->state))
    {
      fprintf(stderr, "Invalid persist file: %s\n", self->persist_filename);
      persist_state_cancel(self->state);
      persist_state_free(self->state);
      self->state = NULL;
      return FALSE;
    }
  return TRUE;
}

void
persist_tool_revert_changes(PersistTool *self)
{
  if (self->state == NULL)
    {
      self->state = self->cfg->state;
    }
  persist_state_cancel(self->state);
  persist_state_start(self->state);
}

PersistTool *
persist_tool_new(gchar *persist_filename, PersistStateMode open_mode)
{
  PersistTool *self = g_new0(PersistTool, 1);
  self->mode = open_mode;
  self->persist_filename = g_strdup(persist_filename);
  self->cfg = cfg_new(0);
  if (!persist_tool_start_state(self))
    {
      persist_tool_free(self);
      return NULL;
    }
  state_handler_register_default_constructors();
  load_state_handler_modules(self->cfg);
  return self;
}

void persist_tool_free(PersistTool *self)
{
  if (self->cfg)
    {
      cfg_free(self->cfg);
    }
  if (self->state)
    {
      if (self->mode != persist_mode_dump)
        {
          persist_state_commit(self->state);
        }
      else
        {
          persist_state_cancel(self->state);
        }
      persist_state_free(self->state);
    }
  g_free(self->persist_filename);
  g_free(self);
}

StateHandler *persist_tool_get_state_handler(PersistTool *self, gchar *name)
{
  STATE_HANDLER_CONSTRUCTOR constructor = NULL;
  StateHandler *handler = NULL;
  gchar *prefix;
  gchar *p;

  p = strchr(name, '(');
  if (!p)
    {
      p = strchr(name, ',');
    }

  if (p)
    {
      prefix = g_strndup(name, p - name);
    }
  else
    {
      prefix = g_strdup(name);
    }
  constructor = state_handler_get_constructor_by_prefix(prefix);
  if (!constructor)
    {
      return NULL;
    }
  handler = constructor(self->state, name);
  return handler;
}

static GOptionEntry dump_options[] =
  {
    { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
  };

static GOptionEntry add_options[] =
  {
      { "output-dir", 'o', 0, G_OPTION_ARG_STRING, &persist_state_dir,"The directory where persist file is located. The name of the persist file stored in this directory must be syslog-ng.persist", "<directory>" },
      { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
  };

static GOptionEntry generate_options[] =
  {
    { "force", 'f', G_OPTION_FLAG_OPTIONAL_ARG, G_OPTION_ARG_NONE, &force_generate, "Overwrite the persist file if it already exists. WARNING: Use this option with care, persist-tool will not ask for confirmation", NULL},
    { "config-file", 'c', 0, G_OPTION_ARG_FILENAME, &config_file_generate, "The syslog-ng configuration file that will be the base of the generated persist", "<config_file>" },
    { "output-dir", 'o', 0, G_OPTION_ARG_FILENAME, &generate_output_dir, "The directory where persist file will be generated to. The name of the persist file will be syslog-ng.persist", "<directory>"},
    { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
  };

static GOptionEntry rename_options[] =
  {
    { "old_key", 'o', 0, G_OPTION_ARG_STRING, &old_key, "Old key in persist file", NULL },
    { "new_key", 'n', 0, G_OPTION_ARG_STRING, &new_key, "New key in persist file", NULL },
    { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
  };

#ifdef _WIN32
static GOptionEntry generate_agent_options[] =
  {
    { "force", 'f', G_OPTION_FLAG_OPTIONAL_ARG, G_OPTION_ARG_NONE, &force_generate_agent, "Overwrite the persist file if it already exists. WARNING: Use this option with care, persist-tool will not ask for confirmation", NULL},
    { "output-dir", 'o', 0, G_OPTION_ARG_FILENAME, &generate_agent_output_dir, "The directory where persist file will be generated to. The name of the persist file will be syslog-ng.persist", "<directory>"},
    { "xml-config", 'x', G_OPTION_FLAG_OPTIONAL_ARG, G_OPTION_ARG_FILENAME, &xml_config_file, "The XML configuration file of the syslog-ng Agent for Windows application. If this parameter is set, the persist file will be generated from the XML configuration file. If this parameter is missing, the configuration available in the system registry will be used.", NULL},
    { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
  };
#endif




typedef struct _PersistToolMode
{
  const gchar *mode;
  const GOptionEntry *options;
  const gchar *description;
  const gchar *parameter_description;
  gint
  (*main)(gint argc, gchar *argv[]);
} PersistToolMode;

#ifdef _WIN32
static PersistToolMode modes[] =
  {
    { "dump", dump_options, "Dump the contents of the persist file", "<persist_file>", dump_main },
    { "rename-key", rename_options, "Rename a key in the persist file with new key", "<persist_file>", rename_main },
    { "generate", generate_options, "Generate a persist file from the configuration file of syslog-ng", "", generate_main },
    { "generate-agent", generate_agent_options,
       "Generate persist file for the syslog-ng Agent for Windows application", "", generate_agent_main },

    { "add", add_options, "Add new or change entry in the persist file", "<input_file>", add_main },
    { NULL, NULL }, };
#else
static PersistToolMode modes[] =
  {
    { "dump", dump_options, "Dump the contents of the persist file", "<persist_file>", dump_main },
    { "rename-key", rename_options, "Rename a key in the persist file with new key", "<persist_file>", rename_main },
    { "generate", generate_options, "Generate a persist file from the configuration file of syslog-ng", "", generate_main },
    { "add", add_options, "Add new or change entry in the persist file", "<input_file>", add_main },
    { NULL, NULL }, };
#endif

const gchar *
get_mode(int *argc, char **argv[])
{
  gint i;
  const gchar *mode;

  for (i = 1; i < (*argc); i++)
    {
      if ((*argv)[i][0] != '-')
        {
          mode = (*argv)[i];
          memmove(&(*argv)[i], &(*argv)[i+1], ((*argc) - i) * sizeof(gchar *));
          (*argc)--;
          return mode;
        }
    }
  return NULL;
}

void
usage(const gchar *bin_name)
{
  gint mode;

  fprintf(stderr, "Syntax: %s <command> [options]\nPossible commands are:\n",
      bin_name);
  for (mode = 0; modes[mode].mode; mode++)
    {
      fprintf(stderr, "    %-20s %s\n", modes[mode].mode,
          modes[mode].description);
    }
  exit(1);
}

int
main(int argc, char *argv[])
{
  const gchar *mode_string;
  GOptionContext *ctx;
  gint mode;
  GError *error = NULL;
  int result;

  setvbuf(stderr, NULL, _IONBF, 0);
  mode_string = get_mode(&argc, &argv);
  if (!mode_string)
    {
      usage(argv[0]);
    }

  ctx = NULL;

  for (mode = 0; modes[mode].mode; mode++)
    {
      if (strcmp(modes[mode].mode, mode_string) == 0)
        {
          ctx = g_option_context_new(modes[mode].parameter_description);
#if GLIB_CHECK_VERSION (2, 12, 0)
          g_option_context_set_summary(ctx, modes[mode].description);
#endif
          g_option_context_add_main_entries(ctx, modes[mode].options, NULL );
          break;
        }
    }
  if (!ctx)
    {
      fprintf(stderr, "Unknown command\n");
      usage(argv[0]);
    }

  argv[0] = g_strdup_printf("%s %s", argv[0], mode_string);

  if (!g_option_context_parse(ctx, &argc, &argv, &error))
    {
      fprintf(stderr, "Error parsing command line arguments: %s\n",
          error ? error->message : "Invalid arguments");
      g_clear_error(&error);
      g_option_context_free(ctx);
      return 1;
    }
  g_option_context_free(ctx);

  result = modes[mode].main(argc, argv);
  return result;
}
